### PR TITLE
[simantak-dabhade] iP

### PR DESCRIPTION
**Pull request by simantak-dabhade**
# 🏀 LEBROOOOONNNN Task Manager - Complete Feature Implementation

## Summary

This PR implements a comprehensive task management chatbot named "Lebron" - your personal GOAT assistant for staying organized! The implementation includes a full-featured CLI application with persistent storage, intelligent command parsing, and robust error handling.

### Core Task Management
- **Todo Tasks**: Simple task tracking with completion status
- **Deadline Tasks**: Tasks with due dates and smart date parsing
- **Event Tasks**: Time-bounded tasks with start and end times
- **Task Operations**: Mark/unmark as complete, delete tasks, list all tasks

### Advanced Features
- **🔍 Find Command**: Search tasks by keywords
- **💾 Persistent Storage**: Automatic saving/loading from file system
- **🎯 Smart Command Parsing**: Robust input validation and error handling
- **📝 Comprehensive JavaDoc**: Detailed documentation throughout codebase

### Architecture Highlights
- **MVC Pattern**: Clean separation with UI, Storage, Parser, and Data layers
- **Package Organization**: Well-structured `lebron.*` package hierarchy
- **Error Handling**: Graceful degradation with user-friendly error messages
- **Testing Suite**: Unit tests for critical components (Parser, TaskList)

### Command Processing
- Implements command pattern for extensible operation handling
- Supports commands: `todo`, `deadline`, `event`, `list`, `mark`, `unmark`, `delete`, `find`, `bye`
- Robust input validation with helpful error messages

### Data Persistence
- Automatic file-based storage in `./data/lebron_data.txt`
- Graceful handling of corrupted or missing data files
- Efficient serialization/deserialization

### Date Handling
- Smart date parsing for deadlines and events
- Flexible input formats supported
- Validation for logical date ranges

## 🧪 Testing

- **Unit Tests**: Comprehensive test coverage for Parser and TaskList components
- **Manual Testing**: CLI interface thoroughly tested with edge cases
- **Error Scenarios**: Validated graceful handling of invalid inputs

## 🎯 Usage Examples

```bash
# Add a simple todo
todo read book

# Add a deadline
deadline submit assignment /by 2024-12-31

# Add an event
event team meeting /from 2024-12-15 10:00 /to 2024-12-15 11:00

# Find tasks containing "book"
find book

# Mark task as complete
mark 1

# List all tasks
list
```

## 🔧 Build & Run

```bash
# Build the project
./gradlew build

# Run the application
./gradlew run

# Run tests
./gradlew test
```

*Just like the le pookie, le GOAT himself, this task manager delivers reliable performance when you need it most!* 🐐

### CODE REVIEW

File: src/main/java/lebron/Lebron.java [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+ * Features include:
+ * - Managing different types of tasks (todos, deadlines, events)
+ * - Persistent storage of your tasks
+ * - Smart date parsing for deadlines and events
+ * - Search functionality to find specific tasks
+ * 
+ * @author Your friendly neighborhood developer
+ * @version 1.0
+ */
+public class Lebron {
```
Comment (gab-er) [inline src/main/java/lebron/Lebron.java:26]: I like the use of the Javadoc comments here to briefly explain more about the chatbot

File: src/main/java/lebron/Lebron.java [hunk 2]
```diff
@@ -0,0 +1,291 @@
+package lebron;
+
+import lebron.storage.Storage;
+import lebron.data.TaskList;
+import lebron.ui.Ui;
+import lebron.parser.Parser;
+import lebron.task.*;
+import java.util.List;
```
Comment (gab-er) [inline src/main/java/lebron/Lebron.java:9]: I like the organisation of package and imports

File: src/main/java/lebron/Lebron.java [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+    }
+
+    /**
+     * Starts the main chat loop where all the magic happens!
+     * 
+     * This method handles the interactive conversation with you, processing each command
+     * and keeping things running smoothly until you decide to say goodbye. It's designed
+     * to be forgiving - if something goes wrong, we'll catch it and keep going.
+     */
+    public void run() {
```
Comment (gab-er) [inline src/main/java/lebron/Lebron.java:57]: The function code looks really clean and the code is easy to read

File: src/test/java/lebron/data/TaskListTest.java [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+import lebron.task.Deadline;
+import lebron.task.Event;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.*;
+
+import java.util.ArrayList;
+import java.util.List;
+
+class TaskListTest {
```
Comment (gab-er) [inline src/test/java/lebron/data/TaskListTest.java:14]: I like the wide range of testing done and the naming convention is clear, making it easy to understand what each test does

File: src/main/java/lebron/storage/Storage.java [hunk 5]
```diff
@@ -0,0 +1,145 @@
+package lebron.storage;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.io.*;
```
Comment (gab-er) [inline src/main/java/lebron/storage/Storage.java:5]: I believe that importing the whole package like this using * should be avoided as much as possible, so maybe it is best to import the specific module you need instead

**Review by gab-er — COMMENTED** [review]
Overall, the code looks really good and the javadoc comments are clear. Nice!

**Comment by cameronlzy** [comment]
I really like ur new changes well done simantak!

File: src/main/java/lebron/Lebron.java [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+    private void handleDeadline(String input) {
+        if (input.trim().isEmpty()) {
+            ui.showError("The description of a deadline cannot be empty.\\nPlease use format: deadline <description> /by <date>");
+            return;
+        }
+        
+        String[] parts = input.split(" /by ", 2);
+        if (parts.length == 2) {
+            String description = parts[0].trim();
+            String by = parts[1].trim();
```
Comment (Ma-Yirui) [inline src/main/java/lebron/Lebron.java:135]: Consider changing the name of variable to a noun, maybe "byTime" ?

File: src/main/java/lebron/Lebron.java [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+            return;
+        }
+        
+        String[] parts = input.split(" /from ", 2);
+        if (parts.length == 2) {
+            String description = parts[0].trim();
+            String[] timeParts = parts[1].split(" /to ", 2);
+            if (timeParts.length == 2) {
+                String from = timeParts[0].trim();
+                String to = timeParts[1].trim();
```
Comment (Ma-Yirui) [inline src/main/java/lebron/Lebron.java:172]: similar to the comment above, consider using nouns for variable names.

File: src/main/java/lebron/parser/Parser.java [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+     * 
+     * @param input what you typed in the chat
+     * @return a Command object with the parsed command type and arguments
+     */
+    public static Command parse(String input) {
+        if (input == null || input.trim().isEmpty()) {
+            return new Command(CommandType.UNKNOWN, "");
+        }
+
+        String trimmed = input.trim();
```
Comment (Ma-Yirui) [inline src/main/java/lebron/parser/Parser.java:75]:
```suggestion
        String trimmedInput = input.trim();
```

File: src/main/java/lebron/task/Deadline.java [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+    public String getByString() {
+        return by.format(OUTPUT_FORMATTER);
+    }
+
+    private LocalDateTime parseDateTime(String dateTimeStr) {
+        if (dateTimeStr == null || dateTimeStr.trim().isEmpty()) {
+            throw new IllegalArgumentException("Date cannot be empty");
+        }
+
+        String trimmed = dateTimeStr.trim();
```
Comment (Ma-Yirui) [inline src/main/java/lebron/task/Deadline.java:77]:
```suggestion
        String trimmedInput = input.trim();
```

File: src/main/java/lebron/task/Event.java [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+    public String getToString() {
+        return to.format(OUTPUT_FORMATTER);
+    }
+
+    private LocalDateTime parseDateTime(String dateTimeStr) {
+        if (dateTimeStr == null || dateTimeStr.trim().isEmpty()) {
+            throw new IllegalArgumentException("Date cannot be empty");
+        }
+
+        String trimmed = dateTimeStr.trim();
```
Comment (Ma-Yirui) [inline src/main/java/lebron/task/Event.java:100]:
```suggestion
        String trimmedInput = input.trim();
```

File: src/main/java/lebron/task/Task.java [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+/**
+ * The base class for all types of tasks in your list.
+ * 
+ * Whether it's a simple todo, a deadline, or an event, they all share some common
+ * characteristics - they have a description, they can be marked as done, and they
+ * have a visual representation. This abstract class captures all that shared behavior.
+ */
+public abstract class Task {
+    protected String description;
+    protected boolean done;
```
Comment (Ma-Yirui) [inline src/main/java/lebron/task/Task.java:12]: Booleans should sound like booleans.
Consider using "isDone" or similar names.

**Review by Ma-Yirui — COMMENTED** [review]
Overall good naming practices. Good Job!
